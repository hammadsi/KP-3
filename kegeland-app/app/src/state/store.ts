import {combineReducers, configureStore} from '@reduxjs/toolkit';
import AsyncStorage from '@react-native-async-storage/async-storage';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import reduxFlipper from 'redux-flipper';

import appReducer from './ducks/app/app.reducer';
import bluetoothReducer from './ducks/bluetooth/bluetooth.reducer';
import authReducer from './ducks/auth/auth.reducer';
import questionsReducer from './ducks/questions/questions.reducer';
import sessionReducer from './ducks/session/session.reducer';

import rootSaga, {sagaMiddleware} from './rootSaga';

/**
 * Configuration for redux-persist.
 * Redux-persist is responsible for maintaining the state
 * between sessions.
 */
const persistConfig = {
  key: 'root',
  version: 1,
  storage: AsyncStorage,
  whitelist: ['app', 'auth', 'bluetooth'],
};

// Initiate the root reducer
export const rootReducer = combineReducers({
  app: appReducer,
  auth: authReducer,
  bluetooth: bluetoothReducer,
  questions: questionsReducer,
  session: sessionReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

// Check if development-environment is enabled
const isDev = !['production', 'test', 'integration'].includes(
  process.env.NODE_ENV || '',
);

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        })
          .concat(logger)
          .concat(reduxFlipper())
          .concat(sagaMiddleware)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(sagaMiddleware),
  devTools: isDev,
});

// Start redux-saga
sagaMiddleware.run(rootSaga);

// Create persistor for persisting state
export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
