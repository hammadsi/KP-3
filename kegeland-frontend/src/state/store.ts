import { combineReducers, configureStore } from '@reduxjs/toolkit';
import logger from 'redux-logger';
import {
  persistStore,
  persistReducer,
  FLUSH,
  REHYDRATE,
  PAUSE,
  PERSIST,
  PURGE,
  REGISTER,
} from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import authReducer from './ducks/auth/auth.reducer';
import patientsReducer from './ducks/patients/patients.reducer';
import questionnairesReducer from './ducks/questionnaires/questionnaires.reducer';
import { sensorReducer } from './ducks/sensors/sensors.reducer';
import sessionsReducer from './ducks/sessions/sessions.reducer';
import settingsReducer from './ducks/settings/settings.reducer';
import wheelchairPatientsReducer from './ducks/wheelchairPatients/wheelchairPatients.reducer';

const persistConfig = {
  key: 'root',
  version: 1,
  storage,
  whitelist: ['settings', 'auth'],
};

export const rootReducer = combineReducers({
  settings: settingsReducer,
  auth: authReducer,
  patients: patientsReducer,
  sensors: sensorReducer,
  sessions: sessionsReducer,
  questionnaires: questionnairesReducer,
  wheelchairPatients: wheelchairPatientsReducer,
});

const persistedReducer = persistReducer(persistConfig, rootReducer);

const isDev = !['production', 'test'].includes(process.env.NODE_ENV || '');

export const store = configureStore({
  reducer: persistedReducer,
  middleware: (getDefaultMiddleware) =>
    isDev
      ? getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }).concat(logger)
      : getDefaultMiddleware({
          serializableCheck: {
            ignoredActions: [FLUSH, REHYDRATE, PAUSE, PERSIST, PURGE, REGISTER],
          },
        }),
  devTools: isDev,
});

export const persistor = persistStore(store);

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
