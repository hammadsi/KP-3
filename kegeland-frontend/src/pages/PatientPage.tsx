import React, { useState } from 'react';
import {
  Button,
  Box,
  Flex,
  Stack,
  useMediaQuery,
  Center,
} from '@chakra-ui/react';

import { useParams } from 'react-router-dom';
import {
  AiOutlineClockCircle,
  AiOutlineStock,
  AiOutlineCalendar,
} from 'react-icons/ai';
import { useSelector } from 'react-redux';

import Card from '../components/Card';
import LabeledValue from '../components/LabeledValue';
import withLayout from '../hoc/withLayout';
import usePatient from '../hooks/usePatient';
import WeeklySessionsChart from '../components/WeeklySessionsChart';
import withSpinner from '../hoc/withSpinner';
import useAppSelector from '../hooks/useAppSelector';
import { UserRole } from '../state/ducks/auth/auth.interface';
import FemfitExerciseTable from '../components/FemfitExerciseTable';
import WheelchairExerciseTable from '../components/WheelchairExerciseTable';
import { RootState } from '../state/store';
import useWheelchairPatient from '../hooks/useWheelchairPatient';
import { ViewSession } from '../state/ducks/sessions/sessions.interface';
import useUploadIMUData from '../hooks/useUploadIMUData';

type PatientPageParams = {
  patientId: string;
};

const PatientPage: React.FC = () => {
  const [isGreaterThanLg] = useMediaQuery('(min-width: 62em)');
  const { patientId } = useParams<PatientPageParams>();
  const { data, details: femfitDetails, loading } = usePatient(patientId || '');
  const { authUser } = useSelector((state: RootState) => state.auth);
  const { gameSessions, details: wheelchairDetails } = useWheelchairPatient(
    authUser?.id,
  );
  const { userDetails } = useAppSelector((state) => state.auth);
  const { uploadIMUData, loading: imuUploadLoading, error: imuUploadError } = useUploadIMUData();

  const sortedGameSessions = [...gameSessions].sort(
    (a, b) => b.createdAt - a.createdAt,
  );
  const sortedData = [...data].sort((a, b) => b.createdAt - a.createdAt);

  const allSessions: ViewSession[] = [...sortedData, ...sortedGameSessions];

  const headingStyle = {
    color: 'var(--chakra-colors-blackAlpha-800)',
    fontWeight: 'bold',
    fontSize: '24px',
    margin: '25px 0 10px 0',
  };

  // New code for file upload
  const [selectedFile, setSelectedFile] = useState<File | null>(null);
  const [uploadStatus, setUploadStatus] = useState<'idle' | 'done'>('idle');

  const handleFileSelection = (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files?.[0];
    if (file) {
      setSelectedFile(file);
    }
  };

  const triggerFileUpload = () => {
    if (selectedFile) {
      const reader = new FileReader();
      reader.onload = async function (event) {
        const text = event?.target?.result?.toString();
        if (!text) {
          console.error('Could not read file.');
          return;
        }
        const lines = text.split('\n');
        const imuData = [];

        for (let i = 1; i < lines.length; i++) {
          const line = lines[i];
          const parts = line.split(',');

          if (parts.length === 7) {
            const [
              timestamp,
              x_accel,
              y_accel,
              z_accel,
              x_gyro,
              y_gyro,
              z_gyro,
            ] = parts.map(part => parseFloat(part));

            imuData.push({
              timestamp: timestamp,
              accelerometer: {
                x: x_accel,
                y: y_accel,
                z: z_accel,
              },
              gyroscope: {
                x: x_gyro,
                y: y_gyro,
                z: z_gyro,
              },
            });
          }
        }

        console.log('Parsed IMU data:', imuData);
        try {
          const patientId = 'Wwy4sqcl7dYGvvkHA5mmdWBEa713';
          const sessionId = 'faDFwwohudvgI5GjBsM9';

          await uploadIMUData(patientId, sessionId, imuData);
          setUploadStatus('done');
        } catch (error) {
          console.error('Error uploading IMU data:', error);
          setUploadStatus('idle');
        }
        // TODO: Update the Firestore with Update calls to the API when endpoints are ready.
        setUploadStatus('done');
      };

      reader.onerror = function (event) {
        console.error('An error occurred while reading the file:', event);
        setUploadStatus('idle');
      };

      reader.readAsText(selectedFile);
    }
  };

  const startUnitySession = () => {
    window.location.href = `VRWheelchairSim:// -patientID ${patientId} -bearerToken ${localStorage.getItem(
      'id_token',
    )}`;
  };

  return (
    <Box w="100%" h="100%">
      <Flex
        flexDirection={isGreaterThanLg ? 'row' : 'column'}
        flexBasis="100%"
        flexWrap="nowrap">
        <Card
          marginRight={5}
          w={isGreaterThanLg ? '25%' : '100%'}
          minH={isGreaterThanLg ? 'md' : undefined}
          loading={loading}>
          <Stack
            spacing={4}
            direction={isGreaterThanLg ? 'column' : 'row'}
            w="100%"
            alignItems="flex-start">
            <LabeledValue
              label="Workouts this week"
              value={
                femfitDetails.sessionsThisWeek +
                wheelchairDetails.sessionsThisWeek
              }
              icon={AiOutlineCalendar}
            />
            {/* TODO: Only shpw the one relevant for patient type */}
            <LabeledValue
              label="Since last femfit exercise"
              value={femfitDetails.lastSessionDelta}
              icon={AiOutlineClockCircle}
            />
            <LabeledValue
              label="Since last wheelchair exercise"
              value={wheelchairDetails.lastSessionDelta}
              icon={AiOutlineClockCircle}
            />
            <LabeledValue
              label="Total workouts"
              value={
                femfitDetails.sessionsTotal + wheelchairDetails.sessionsTotal
              }
              icon={AiOutlineStock}
            />
          </Stack>
        </Card>
        <Card
          w={isGreaterThanLg ? '75%' : '100%'}
          minH={isGreaterThanLg ? 'md' : undefined}
          loading={loading}>
          <WeeklySessionsChart sessions={allSessions} numWeeks={12} />
        </Card>
      </Flex>

      {/* TODO: Only show the ones relevant for the patient type */}

      <h1 style={headingStyle}>Overview of Femfit exercises</h1>
      <Card loading={loading} minH="36">
        <FemfitExerciseTable sessions={sortedData} patientId={patientId!} />
      </Card>

      <h1 style={headingStyle}>Overview of Wheelchair exercises</h1>
      <Card loading={loading} minH="36">
        <WheelchairExerciseTable
          sessions={sortedGameSessions}
          patientId={patientId!}
        />
      </Card>

      {userDetails?.roles.includes(UserRole.PHYSICIAN) && (
        <Button w="100%" marginTop={8} onClick={startUnitySession}>
          Start session
        </Button>
      )}


      {/*Move this into the Exercise Session Page when the issue is done */}
      <Card>
        <h2 style={headingStyle}> Upload IMU data for session</h2>
        <Center marginTop={12}>
          <input
            type="file"
            id="imuData"
            name="imuData"
            accept=".csv"
            onChange={handleFileSelection}
          />
        <Button
          onClick={triggerFileUpload}
          ml={4}
          isDisabled={!selectedFile || imuUploadLoading}
          colorScheme={uploadStatus === 'done' ? 'green' : 'blue'}>
          {uploadStatus === 'done' ? 'Uploaded' : imuUploadLoading ? 'Uploading...' : 'Upload selected IMU Data'}
        </Button>
        {imuUploadError && <p>Error uploading data: {imuUploadError}</p>}
        </Center>
      </Card>
    </Box>
  );
};

export default withLayout(withSpinner(PatientPage));
