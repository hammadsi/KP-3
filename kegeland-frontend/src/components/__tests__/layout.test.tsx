import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import { mockStore } from '../../state/mocks/store.mock';
import PatientPage from '../../pages/PatientPage';

jest.mock('../../hooks/useSilentRefresh');
jest.mock('../../hooks/useBreadcrumbs', () => {
  return [];
});
describe('Test layout', () => {
  it('renders correctly', () => {
    const store = mockStore();
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <PatientPage />
          </MemoryRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
