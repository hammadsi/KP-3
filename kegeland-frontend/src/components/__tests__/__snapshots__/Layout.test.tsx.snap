// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test layout renders correctly 1`] = `null`;
